lib/eth/abi/type.rb

```diff
@@ -122,7 +131,21 @@ def is_dynamic?
       #
       # @return [Eth::Abi::Type] nested sub-type.
       def nested_sub
-        @nested_sub ||= self.class.new(base_type, sub_type, dimensions[0...-1])
+        @nested_sub ||= self.class.new(base_type, sub_type, dimensions[0...-1], components, name)
+      end
+
+      def to_s
+        if base_type == "tuple"
+          "(" + components.map(&:to_s).join(",") + ")" + (dimensions.size > 0 ? dimensions.map { |x| "[#{x == 0 ? "" : x}]" }.join : "")
+        elsif dimensions.empty?
+          if %w[string bytes].include?(base_type) and sub_type.empty?
```

Comment:
- [`and`/`or` behaves differently than `&&`/`||`](https://avdi.codes/using-and-and-or-in-ruby/)
- [`and`/`or` is misunderstood and causes hard to find bugs](https://bulldogjob.com/readme/ruby-gotchas-that-will-come-back-to-haunt-you#:~:text=%E2%80%9Cand%E2%80%9D%20is%20NOT%20the%20same%20as%C2%A0%E2%80%9C%26%26%E2%80%9D)
- [`and`/`or` is usually just strictly forbidden from many ruby style guides](https://github.com/airbnb/ruby#no-and-or)

```suggestion
          if %w[string bytes].include?(base_type) && sub_type.empty?
```